drivers/spi/spi_ambiq_spis.c

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright (c) 2024 Ambiq <www.ambiq.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ambiq_spis
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(spi_ambiq_spis);
+
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/policy.h>
+#include <zephyr/pm/device_runtime.h>
+
+#include <stdlib.h>
+#include <errno.h>
+#include "spi_context.h"
+#include <am_mcu_apollo.h>
+
+#define PWRCTRL_MAX_WAIT_US 5
+
+typedef int (*ambiq_spi_pwr_func_t)(void);
+
+struct spi_ambiq_config {
+	const struct gpio_dt_spec int_gpios;
+	uint32_t base;
+	int size;
+	const struct pinctrl_dev_config *pcfg;
+	ambiq_spi_pwr_func_t pwr_func;
+	void (*irq_config_func)(void);
+};
+
+struct spi_ambiq_data {
+	struct spi_context ctx;
+	am_hal_ios_config_t ios_cfg;
+	void *ios_handler;
+	int inst_idx;
+	struct k_sem spim_wrcmp_sem;
+};
+
+#define AMBIQ_SPIS_TX_BUFSIZE_MAX 1023
+uint8_t g_pui8TxFifoBuffer[AMBIQ_SPIS_TX_BUFSIZE_MAX];
+
+#define SPI_WORD_SIZE 8
+
+#define FIFO_BASE   0x78
+#define FIFO_END    0x100
+#define FIFO_LENGTH (FIFO_END - FIFO_BASE)
+
+#define AMBIQ_SPIS_INT_ERR (AM_HAL_IOS_INT_FOVFL | AM_HAL_IOS_INT_FUNDFL | AM_HAL_IOS_INT_FRDERR)
+
+#define AMBIQ_SPIS_XCMP_INT                                                                        \
+	(AM_HAL_IOS_INT_XCMPWR | AM_HAL_IOS_INT_XCMPWF | AM_HAL_IOS_INT_XCMPRR |                   \
+	 AM_HAL_IOS_INT_XCMPRF)
+
+static void spi_ambiq_reset(const struct device *dev)
+{
+	struct spi_ambiq_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+
+	/* cancel timed out transaction */
+	am_hal_ios_disable(data->ios_handler);
+	/* NULL config to trigger reconfigure on next xfer */
+	ctx->config = NULL;
+	/* signal any thread waiting on sync semaphore */
+	spi_context_complete(ctx, dev, -ETIMEDOUT);
+	/* clean up for next xfer */
+	k_sem_reset(&ctx->sync);
```

Comment:
You should not have to do that. The sync lock cannot be taken more than once, so after spi_context_complete: it will be available anyway.